// vxe-table/lib/radio/src/group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _conf = _interopRequireDefault(require("../../v-x-e-table/src/conf"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default2 = exports.default = {
  name: 'VxeRadioGroup',
  props: {
    value: [String, Number, Boolean],
    type: String,
    options: Array,
    optionProps: Object,
    disabled: Boolean,
    strict: {
      type: Boolean,
      default: function _default() {
        return _conf.default.radioGroup.strict;
      }
    },
    size: {
      type: String,
      default: function _default() {
        return _conf.default.radioGroup.size || _conf.default.size;
      }
    }
  },
  inject: {
    $xeform: {
      default: null
    },
    $xeformiteminfo: {
      default: null
    }
  },
  provide: function provide() {
    return {
      $xeradiogroup: this
    };
  },
  computed: {
    vSize: function vSize() {
      return this.size || this.$parent.size || this.$parent.vSize;
    },
    propsOpts: function propsOpts() {
      return this.optionProps || {};
    },
    labelField: function labelField() {
      return this.propsOpts.label || 'label';
    },
    valueField: function valueField() {
      return this.propsOpts.value || 'value';
    },
    disabledField: function disabledField() {
      return this.propsOpts.disabled || 'disabled';
    }
  },
  data: function data() {
    return {
      name: _xeUtils.default.uniqueId('xegroup_')
    };
  },
  render: function render(h) {
    var $scopedSlots = this.$scopedSlots,
      options = this.options,
      type = this.type,
      valueField = this.valueField,
      labelField = this.labelField,
      disabledField = this.disabledField;
    var defaultSlot = $scopedSlots.default;
    var btnComp = type === 'button' ? 'vxe-radio-button' : 'vxe-radio';
    return h('div', {
      class: 'vxe-radio-group'
    }, defaultSlot ? defaultSlot.call(this, {}) : options ? options.map(function (item) {
      return h(btnComp, {
        props: {
          label: item[valueField],
          content: item[labelField],
          disabled: item[disabledField]
        }
      });
    }) : []);
  },
  methods: {
    handleChecked: function handleChecked(params, evnt) {
      this.$emit('input', params.label);
      this.$emit('change', params);
      // 自动更新校验状态
      if (this.$xeform && this.$xeformiteminfo) {
        this.$xeform.triggerItemEvent(evnt, this.$xeformiteminfo.itemConfig.field, params.label);
      }
    }
  }
};
