// vxe-table/lib/v-x-e-table/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  use: true,
  setConfig: true,
  globalConfs: true,
  t: true,
  _t: true,
  v: true,
  setup: true,
  config: true,
  setIcon: true,
  globalStore: true,
  getComponent: true,
  component: true,
  version: true,
  tableVersion: true,
  VxeUI: true,
  VXETable: true
};
exports.VxeUI = exports.VXETable = void 0;
exports._t = _t;
exports.component = component;
exports.default = exports.config = void 0;
exports.getComponent = getComponent;
exports.setConfig = exports.globalStore = exports.globalConfs = void 0;
exports.setIcon = setIcon;
exports.setup = void 0;
exports.t = t;
exports.tableVersion = void 0;
exports.use = use;
exports.version = exports.v = void 0;
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _domZindex = _interopRequireDefault(require("dom-zindex"));
var _conf = _interopRequireDefault(require("../v-x-e-table/src/conf"));
var _interceptor = require("./src/interceptor");
Object.keys(_interceptor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interceptor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _interceptor[key];
    }
  });
});
var _renderer = require("./src/renderer");
Object.keys(_renderer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _renderer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _renderer[key];
    }
  });
});
var _commands = require("./src/commands");
Object.keys(_commands).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commands[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _commands[key];
    }
  });
});
var _menus = require("./src/menus");
Object.keys(_menus).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menus[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _menus[key];
    }
  });
});
var _formats = require("./src/formats");
Object.keys(_formats).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formats[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _formats[key];
    }
  });
});
var _validators = require("./src/validators");
var _theme = require("./src/theme");
Object.keys(_theme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _theme[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _theme[key];
    }
  });
});
var _utils = require("../tools/utils");
var _log = require("../tools/log");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var installedPlugins = [];
function use(Plugin, options) {
  /* eslint-disable @typescript-eslint/no-use-before-define */
  if (Plugin && Plugin.install) {
    if (installedPlugins.indexOf(Plugin) === -1) {
      Plugin.install(VxeUI, options);
      installedPlugins.push(Plugin);
    }
  }
  return VxeUI;
}

/**
 * 检测模块的安装顺序是否正确
 */
function reg(key) {
  /* eslint-disable @typescript-eslint/no-use-before-define */
  // 检测安装顺序是否正确
  if (process.env.NODE_ENV === 'development') {
    if (VxeUI.Table) {
      (0, _log.errLog)('vxe.error.useErr', [key]);
    }
  }
  VxeUI["_".concat(key)] = 1;
}

// function getExportOrImpotType (types, flag) {
//   const rest = []
//   XEUtils.objectEach(types, (val, type) => {
//     if (val === 0 || val === flag) {
//       rest.push(type)
//     }
//   })
//   return rest
// }

/**
 * 全局参数设置
 */
var setConfig = exports.setConfig = function setConfig(options) {
  if (options) {
    if (options.theme) {
      (0, _theme.setTheme)(options.theme);
    }
    if (options.zIndex) {
      _domZindex.default.setCurrent(options.zIndex);
    }
    _xeUtils.default.merge(_conf.default, options);
  }
  return VxeUI;
};
var VXETableConfig = /*#__PURE__*/function () {
  function VXETableConfig() {
    _classCallCheck(this, VXETableConfig);
  }
  return _createClass(VXETableConfig, [{
    key: "zIndex",
    get:
    /**
     * 获取当前的 zIndex
     */
    function get() {
      return _utils.UtilTools.getLastZIndex();
    }

    /**
     * 获取下一个 zIndex
     */
  }, {
    key: "nextZIndex",
    get: function get() {
      return _utils.UtilTools.nextZIndex();
    }

    /**
     * 获取所有导出类型
     */
  }, {
    key: "exportTypes",
    get: function get() {
      return _xeUtils.default.keys(_conf.default.table.exportConfig._typeMaps);
    }

    /**
     * 获取所有导入类型
     */
  }, {
    key: "importTypes",
    get: function get() {
      return _xeUtils.default.keys(_conf.default.table.importConfig._typeMaps);
    }
  }]);
}(); // 已废弃
var globalConfs = exports.globalConfs = new VXETableConfig();
function t(key, args) {
  return _conf.default.i18n(key, args);
}
function _t(key, args) {
  return key ? _xeUtils.default.toValueString(_conf.default.translate ? _conf.default.translate(key, args) : key) : '';
}
var v = exports.v = 'v3';

/**
 * 已废弃，请使用 setConfig
 * @deprecated
 */
var setup = exports.setup = function setup(options) {
  if (process.env.NODE_ENV === 'development') {
    (0, _log.warnLog)('vxe.error.delFunc', ['setup', 'setConfig']);
  }
  return setConfig(options);
};

/**
 * 已废弃，请使用 setConfig
 * @deprecated
 */
var config = exports.config = function config(options) {
  if (process.env.NODE_ENV === 'development') {
    (0, _log.warnLog)('vxe.error.delFunc', ['setup', 'setConfig']);
  }
  return setConfig(options);
};
function setIcon(options) {
  if (options) {
    Object.assign(_conf.default.icon, options);
  }
  return VxeUI;
}
var globalStore = exports.globalStore = {};
var components = {};
function getComponent(name) {
  return components[name] || null;
}
function component(comp) {
  if (comp && comp.name) {
    components[comp.name] = comp;
  }
}
var version = exports.version = "3.8.16";
var tableVersion = exports.tableVersion = version;
var VxeUI = exports.VxeUI = {
  v: v,
  version: version,
  tableVersion: tableVersion,
  reg: reg,
  use: use,
  setConfig: setConfig,
  setIcon: setIcon,
  globalStore: globalStore,
  interceptor: _interceptor.interceptor,
  renderer: _renderer.renderer,
  commands: _commands.commands,
  formats: _formats.formats,
  menus: _menus.menus,
  validators: _validators.validators,
  t: t,
  _t: _t,
  setTheme: _theme.setTheme,
  getTheme: _theme.getTheme,
  getComponent: getComponent,
  component: component,
  // 已废弃
  config: config,
  setup: setup,
  globalConfs: globalConfs
};
var VXETable = exports.VXETable = VxeUI;
(0, _theme.setTheme)('light');
var _default = exports.default = VxeUI;
